File: src/main/java/frc/robot/util/TunablePIDController.java
package frc.robot.util;

import edu.wpi.first.math.controller.PIDController;
import edu.wpi.first.wpilibj.shuffleboard.Shuffleboard;
import org.littletonrobotics.junction.LogTable;
import org.littletonrobotics.junction.Logger;
import org.littletonrobotics.junction.inputs.LoggableInputs;

public class TunablePIDController extends PIDController implements LoggableInputs {
    private final String name;

    public TunablePIDController(String name, double kp, double ki, double kd) {
        super(kp, ki, kd);
        this.name = name;

        Shuffleboard.getTab("Tunable PID Controllers").add(name, this);
        process();
    }

    private void process() {
        Logger.processInputs("Inputs/TunablePIDControllers/" + name, this);
    }


    @Override
    public void toLog(LogTable table) {
        table.put("P", getP());
        table.put("I", getI());
        table.put("D", getD());
        table.put("IZone", getIZone());
        table.put("PosTol", getPositionTolerance());
        table.put("VelTol", getVelocityTolerance());
    }

    @Override
    public void fromLog(LogTable table) {
        System.out.printf(
                "TunablePIDControllers do not listen to log changes; this one (%s):\n- P = %d\n- I = %d\n- D = %d\n- IZone = %d\n- PosTol = %d\n- VelTol = %d",
                name,
                table.get("P", -1),
                table.get("I", -1),
                table.get("D", -1),
                table.get("IZone", -1),
                table.get("PosTol", -1),
                table.get("VelTol", -1)
        );
    }

    @Override
    public void setP(double kp) {
        super.setP(kp);
        process();
    }

    @Override
    public void setI(double ki) {
        super.setI(ki);
        process();
    }

    @Override
    public void setD(double kd) {
        super.setD(kd);
        process();
    }

    @Override
    public void setIZone(double iZone) {
        super.setIZone(iZone);
        process();
    }

    @Override
    public void setSetpoint(double setpoint) {
        super.setSetpoint(setpoint);
        process();
    }

    @Override
    public void setTolerance(double positionTolerance) {
        super.setTolerance(positionTolerance);
        process();
    }

    @Override
    public void setTolerance(double positionTolerance, double velocityTolerance) {
        super.setTolerance(positionTolerance, velocityTolerance);
        process();
    }
}

File: src/main/java/frc/robot/util/TunablePIDController.kt
package frc.robot.util

import edu.wpi.first.math.controller.PIDController
import edu.wpi.first.wpilibj.shuffleboard.Shuffleboard
import org.littletonrobotics.junction.LogTable
import org.littletonrobotics.junction.Logger
import org.littletonrobotics.junction.inputs.LoggableInputs

class TunablePIDController(private val name: String, kp: Double, ki: Double, kd: Double) : PIDController(kp, ki, kd),
    LoggableInputs {
    init {
        Shuffleboard.getTab("Tunable PID Controllers").add(name, this)
        process()
    }

    private fun process() {
        Logger.processInputs("Inputs/TunablePIDControllers/$name", this)
    }


    override fun toLog(table: LogTable) {
        table.put("P", p)
        table.put("I", i)
        table.put("D", d)
        table.put("IZone", iZone)
        table.put("PosTol", positionTolerance)
        table.put("VelTol", velocityTolerance)
    }

    override fun fromLog(table: LogTable) {
        System.out.printf(
            "TunablePIDControllers do not listen to log changes; values for %s are as follows:\n- P = %d\n- I = %d\n- D = %d\n- IZone = %d\n- PosTol = %d\n- VelTol = %d",
            name,
            table["P", -1],
            table["I", -1],
            table["D", -1],
            table["IZone", -1],
            table["PosTol", -1],
            table["VelTol", -1]
        )
    }

    override fun setP(kp: Double) {
        super.setP(kp)
        process()
    }

    override fun setI(ki: Double) {
        super.setI(ki)
        process()
    }

    override fun setD(kd: Double) {
        super.setD(kd)
        process()
    }

    override fun setIZone(iZone: Double) {
        super.setIZone(iZone)
        process()
    }

    override fun setSetpoint(setpoint: Double) {
        super.setSetpoint(setpoint)
        process()
    }

    override fun setTolerance(positionTolerance: Double) {
        super.setTolerance(positionTolerance)
        process()
    }

    override fun setTolerance(positionTolerance: Double, velocityTolerance: Double) {
        super.setTolerance(positionTolerance, velocityTolerance)
        process()
    }
}
